README.md

[![Travis-CI Build Status](https://travis-ci.org/geanders/countytimezones.svg?branch=master)](https://travis-ci.org/geanders/countytimezones)

<!-- README.md is generated from README.Rmd. Please edit that file -->
Aims of the package
-------------------

This package allows you to convert time from Coordinated Universal Time (UTC, also known as Zulu Time) to local time for US counties, based on each county's FIPs code. You can use this to convert time zones in dataframes where all observations are from one county, or also in a dataframe where observations come from from many different counties.

Many observational datasets express date-times using UTC, to ensure consistency across time zones. Examples include datasets for satellite-based or hourly weather data and aviation data (Morris 2008). It is often helpful to be able to convert date-time values for observations from UTC to local time, especially when combining with data measured or aggregated using local time (e.g., daily counts of health outcomes). This package allows you to make that conversion in a dataset, as long as each observation is associated with a US county Federal Information Processing Standard (FIPS) code. County time zone designations are based on Olson/IANA time zone conventions (see `?OlsonNames` for more information on these conventions).

These local time conversions take into account whether or not the county was observing Daylight Savings Time at the time of the observation. This package can be used both with datasets where all observations are from the same county and for datasets where observations are associated with a number of different counties.

Some counties include multiple time zones. For these counties, the package calculates local time based on the time zone used in the majority of the county, based on land area. Because some counties have more than one time zone, if you are aggregating data from monitors across a county, you should keep observations in UTC until all aggregating is complete and then convert the final, aggregated value to local time, rather than converting to local time before aggregation.

Accessing the package
---------------------

A stable version of this package is available on CRAN and can be installed directly from there:

``` r
install.packages("countytimezones")
```

The lastest development version of the package can also be loaded directly from GitHub using the `devtools` package:

``` r
library(devtools)
install_github("geanders/countytimezones")
library(countytimezones)
```

Simple example
--------------

As a very simple example, here is how you can calculate local time for a single observation in a single county:

``` r
calc_local_time(date_time = "1999-01-01 08:00", fips = "36061")
#>         local_time local_date         local_tz
#> 1 1999-01-01 03:00 1999-01-01 America/New_York
```

The column of local times is given in a character class because, while this package allows you to convert a dataset with observations from many different counties to the local time of each, a date-time vector in R can only have a single time zone associated with all values.

To calculate the same time in several different counties, you can run:

``` r
ex_fips <- c("36061", "17031", "06037")
calc_local_time(date_time = "1999-01-01 08:00", fips = ex_fips)
#>         local_time local_date            local_tz
#> 1 1999-01-01 03:00 1999-01-01    America/New_York
#> 2 1999-01-01 02:00 1999-01-01     America/Chicago
#> 3 1999-01-01 00:00 1999-01-01 America/Los_Angeles
```

You can also do conversions for more complex datasets, with different date-times and counties for each observation:

``` r
ex_datetime <- c("1999-01-01 08:00", "1999-01-01 09:00",
                 "1999-01-01 10:00")
ex_fips <- c("36061", "17031", "06037")
calc_local_time(date_time = ex_datetime, fips = ex_fips)
#>         local_time local_date            local_tz
#> 1 1999-01-01 03:00 1999-01-01    America/New_York
#> 2 1999-01-01 03:00 1999-01-01     America/Chicago
#> 3 1999-01-01 02:00 1999-01-01 America/Los_Angeles
```

From these examples, you can see that the `calc_local_time` function calculates a new dataframe with columns for `local_time`, `local_date`, and `local_tz` (this last column can be omitted by using the option `include_tz = FALSE`). If you want to add these columns to a dataframe, rather than generating them as a separate dataframe, you can instead use the function `add_local_time`. For example:

``` r
ex_df <- data.frame(datetime = c("1999-01-01 08:00", "1999-01-01 09:00",
                                 "1999-01-01 10:00"),
                    fips = c("36061", "17031", "06037"))
add_local_time(df = ex_df, fips = ex_df$fips,
              datetime_colname = "datetime")
#>           datetime  fips       local_time local_date            local_tz
#> 1 1999-01-01 08:00 36061 1999-01-01 03:00 1999-01-01    America/New_York
#> 2 1999-01-01 09:00 17031 1999-01-01 03:00 1999-01-01     America/Chicago
#> 3 1999-01-01 10:00 06037 1999-01-01 02:00 1999-01-01 America/Los_Angeles
```

More complex examples
---------------------

If you want to convert 8:30 UTC on Jan. 1, 1999 to local time in counties throughout the US, you could use the collection of FIPS codes from the `county_tz` dataset included with the package and run:

``` r
data(county_tzs)
example_df <- data.frame(fips = county_tzs$fips,
                 datetime = "1999-01-01 08:30")
head(example_df)
#>   fips         datetime
#> 1 1001 1999-01-01 08:30
#> 2 1003 1999-01-01 08:30
#> 3 1005 1999-01-01 08:30
#> 4 1007 1999-01-01 08:30
#> 5 1009 1999-01-01 08:30
#> 6 1011 1999-01-01 08:30

example_df <- add_local_time(df = example_df,
                             fips = example_df$fips,
                             datetime_colname = "datetime")
head(example_df)
#>   fips         datetime       local_time local_date        local_tz
#> 1 1001 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
#> 2 1003 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
#> 3 1005 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
#> 4 1007 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
#> 5 1009 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
#> 6 1011 1999-01-01 08:30 1999-01-01 02:30 1999-01-01 America/Chicago
```

You can use functions from the `choroplethr` package to map the local values:

``` r
library(choroplethr)
library(choroplethrMaps)
library(ggplot2)
library(dplyr)

to_plot <- example_df %>%
  select(fips, local_time) %>%
  mutate(local_time = factor(local_time)) %>%
  dplyr::rename(region = fips, value = local_time)
a <- CountyChoropleth$new(to_plot)
a$ggplot_scale <- scale_fill_brewer(type = "qual", drop = FALSE)
a$render()
```

![](README-unnamed-chunk-10-1.png)

Here is the same map for a date-time during the summer, when many counties use Daylight Savings Time:

``` r
example_df2 <- data.frame(fips = county_tzs$fips,
                 datetime = "1999-07-01 08:30") 
example_df2 <- add_local_time(example_df2, fips = example_df2$fips,
                              datetime_colname = "datetime")

to_plot <- example_df2 %>%
  select(fips, local_time) %>%
  mutate(local_time = factor(local_time)) %>%
  dplyr::rename(region = fips, value = local_time)
a <- CountyChoropleth$new(to_plot)
a$ggplot_scale <- scale_fill_brewer(type = "qual", drop = FALSE)
a$render()
```

![](README-unnamed-chunk-11-1.png)

From comparing these two maps, you can see some of counties that don't follow Daylight Savings Time (e.g., Arizona and parts of Indiana, although Indiana's Daylight Savings Time policies have changed more recently).

As another more complex example, the `floyd` example data that comes with this package has data on the date and time when Hurricane Floyd was closest to eastern US counties:

``` r
data(floyd)
head(floyd)
#>    fips closest_time_utc
#> 1 01001 1999-09-15 21:15
#> 2 01003 1999-09-15 17:30
#> 3 01005 1999-09-15 20:00
#> 4 01007 1999-09-15 22:00
#> 5 01009 1999-09-16 00:15
#> 6 01011 1999-09-15 20:30
```

``` r
eastern_states <- c("alabama", "arkansas", "connecticut", "delaware",
                            "district of columbia", "florida", "georgia", "illinois",
                            "indiana", "iowa", "kansas", "kentucky", "louisiana",
                            "maine", "maryland", "massachusetts", "michigan",
                            "mississippi", "missouri", "new hampshire", "new jersey",
                            "new york", "north carolina", "ohio", "oklahoma",
                            "pennsylvania", "rhode island", "south carolina",
                            "tennessee", "texas", "vermont", "virginia",
                            "west virginia", "wisconsin")

library(lubridate)
to_plot <- select(floyd, fips, closest_time_utc) %>%
  mutate(fips = as.numeric(fips),
         closest_time_utc = ymd_hm(closest_time_utc)) %>%
  mutate(closest_time_utc = format(closest_time_utc, "%Y-%m-%d")) %>%
  dplyr::rename(region = fips, value = closest_time_utc)
a <- CountyChoropleth$new(to_plot)
a$ggplot_scale <- scale_fill_brewer(type = "qual", drop = FALSE)
a$set_zoom(eastern_states)
a$render()
```

![](README-unnamed-chunk-13-1.png)

You can use the `add_local_time` function to convert all the UTC date-times from this dataset to each county's local time:

``` r
floyd <- add_local_time(floyd, fips = floyd$fips,
                        datetime_colname = "closest_time_utc")
head(floyd)
#>    fips closest_time_utc       local_time local_date        local_tz
#> 1 01001 1999-09-15 21:15 1999-09-15 16:15 1999-09-15 America/Chicago
#> 2 01003 1999-09-15 17:30 1999-09-15 12:30 1999-09-15 America/Chicago
#> 3 01005 1999-09-15 20:00 1999-09-15 15:00 1999-09-15 America/Chicago
#> 4 01007 1999-09-15 22:00 1999-09-15 17:00 1999-09-15 America/Chicago
#> 5 01009 1999-09-16 00:15 1999-09-15 19:15 1999-09-15 America/Chicago
#> 6 01011 1999-09-15 20:30 1999-09-15 15:30 1999-09-15 America/Chicago
```

``` r
to_plot <- select(floyd, fips, local_date) %>%
  mutate(fips = as.numeric(fips))%>%
  dplyr::rename(region = fips, value = local_date)
a <- CountyChoropleth$new(to_plot)
a$ggplot_scale <- scale_fill_brewer(type = "qual", drop = FALSE)
a$set_zoom(eastern_states)
a$render()
```

![](README-unnamed-chunk-15-1.png)

References
----------

Morris, Doug. 2008. “Time for the Weather-- Translating Zulu.” *Weatherwise* 61 (3): 32–35. doi:[10.3200/WEWI.61.3.32-35](https://doi.org/10.3200/WEWI.61.3.32-35).


NEWS.md
# countytimezones 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* Refined time zone designations for Alaskan counties and a few other counties.
* Added some basic error checking to `calc_local_time`.




R/data.R
#' County time zone designations
#'
#' A dataframe containing the Olson/IANA time zone designation of each US county,
#' based on the county's Federal Information Processing Standard
#' (FIPS) code. This dataset was put together based on county FIPS codes as of
#' the 2010 US Census, using text from the two websites listed in "Sources" to
#' set the time zone designation for each county.
#'
#' @details Some counties cover more than one time zone designation. For
#' example, parts of Coconino, Navajo, and Apache Counties in Arizona
#' follow Daylight Savings Time, while other parts of the county do not. We
#' tried to designate county-level time zones based on the time zone used in
#' the majority of the county, based on population. When aggregating from
#' multiple monitors in a county, the time zone should be left in UTC until data
#' is aggregated, then the final value can be converted to local time, rather
#' than converting to local time before aggregating across a county.
#'
#' @format A dataframe with 3,143 rows and 4 variables:
#' \describe{
#'   \item{fips}{A numeric vector giving the county's four- or five-digit
#'               Federal Information Processing Standard (FIPS) code}
#'   \item{tz}{A character vector with the county's Olson/IANA time zone
#'             designation}
#'   \item{county}{A character vector giving the county's name}
#'   \item{state}{A character vector giving the county's state}
#' }
#'
#' @source
#'
#' \url{http://www.statoids.com/tus.html}
#'
#' \url{http://efele.net/maps/tz/us/}
#'
#' @seealso \code{\link{OlsonNames}}
"county_tzs"

#' Date-times for Hurricane Floyd
#'
#' A dataframe that gives the date-time, in Coordinated Universal Time (UTC),
#' for when Hurricane Floyd was closest to each county in the Eastern United
#' States.
#'
#' @format: A dataframe with 2,396 rows and 2 variables:
#' \describe{
#'  \item{fips}{A numeric vector giving the county's four- or five-digit
#'               Federal Information Processing Standard (FIPS) code}
#'  \item{closest_time_utc}{The date and time, in UTC, when Hurricane Floyd
#'                          was closest to the county (based on a storm
#'                          track interpolated to 15 minute observations)}
#' }
#'
#' @source
#'
#' https://github.com/geanders/hurricaneexposure/tree/master/data
"floyd"


data-raw/clean_timezones_data.R
## Starting with US Census data
## Using info from http://www.statoids.com/tus.html on which counties are in which
## time zone
## Updated to be based on http://efele.net/maps/tz/us/
find_state_tz <- function(state){
  if(state %in% c("AL", "AR", "IL", "IA", "KS", "LA", "MN", "MS", "MO",
                  "NE", "ND", "OK", "SD", "TN", "TX", "WI")){
    tz <- "America/Chicago"
  } else if(state %in% c("CT", "DE", "DC", "FL", "GA", "KY", "ME",
                         "MD", "MA", "MI", "NH", "NJ", "NY", "NC", "OH",
                         "PA", "RI", "SC", "VT", "VA", "WV")){
    tz <- "America/New_York"
  } else if(state %in% c("CO", "MT", "NM", "UT", "WY")){
    tz <- "America/Denver"
  } else if(state %in% c("CA", "NV", "OR", "WA")){
    tz <- "America/Los_Angeles"
  } else if(state %in% c("AK")){
    tz <- "America/Anchorage"
  } else if(state %in% c("AZ")){
    tz <- "America/Phoenix"
  } else if(state %in% c("HI")){
    tz <- "Pacific/Honolulu"
  } else if(state %in% "ID"){
    tz <- "America/Boise"
  } else if(state %in% "IN"){
    tz <- "America/Indiana/Indianapolis"
  }
  return(tz)
}

library(dplyr)
library(tidyr)
county_df <- read.csv("http://www2.census.gov/geo/docs/reference/codes/files/national_county.txt",
                      header = FALSE, as.is = TRUE,
                      col.names = c("state", "st_fips", "ct_fips", "county", "code")) %>%
  mutate(ct_fips = sprintf("%03d", ct_fips)) %>%
  unite(fips, st_fips, ct_fips, sep = "") %>%
  mutate(fips = as.numeric(fips)) %>%
  filter(!(state %in% c("AS", "GU", "MP", "PR", "UM", "VI")) ) %>%
  mutate(tz = sapply(state, find_state_tz),
         county = tolower(county))

county_df[county_df$state == "AK" &
            county_df$county %in% c("aleutians west census area",
                                    "aleutians east borough"),
          "tz"] <- "America/Adak"
county_df[county_df$state == "AK" &
            county_df$county %in% c("yakutat city and borough"),
          "tz"] <- "America/Yakutat"
county_df[county_df$state == "AK" &
            county_df$county %in% c("skagway municipality",
                                    "hoonah-angoon census area",
                                    "juneau city and borough",
                                    "haines borough"),
          "tz"] <- "America/Juneau"
county_df[county_df$state == "AK" &
            county_df$county %in% c("wrangell city and borough",
                                    "petersburg census area",
                                    "sitka city and borough",
                                    "ketchikan gateway borough",
                                    "prince of wales-hyder census area"),
          "tz"] <- "America/Sitka"
county_df[county_df$state == "AK" &
            county_df$county %in% c("bethel census area",
                                    "nome census area",
                                    "wade hampton census area"),
          "tz"] <- "America/Nome"
county_df[county_df$state == "FL" &
            county_df$county %in% paste(c("bay", "calhoun", "escambia", "gulf",
                                          "holmes", "jackson", "okaloosa",
                                          "santa rosa", "walton", "washington"),
                                        "county"), "tz"] <- "America/Chicago"
county_df[county_df$state == "ID" &
            county_df$county %in% paste(c("benewah", "bonner", "boundary",
                                          "clearwater", "kootenai", "latah",
                                          "lewis", "nez perce", "shoshone",
                                          "idaho"),
                                        "county"), "tz"] <- "America/Los_Angeles"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("crawford"), "county"),
          "tz"] <- "America/Indiana/Marengo"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("daviess", "dubois", "knox",
                                          "martin"), "county"),
          "tz"] <- "America/Indiana/Vincennes"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("perry"), "county"),
          "tz"] <- "America/Indiana/Tell_City"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("pike"), "county"),
          "tz"] <- "America/Indiana/Petersburg"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("starke"), "county"),
          "tz"] <- "America/Indiana/Knox"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("pulaski"), "county"),
          "tz"] <- "America/Indiana/Winamac"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("switzerland"), "county"),
          "tz"] <- "America/Indiana/Vevay"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("clark", "floyd", "harrison"), "county"),
          "tz"] <- "America/Kentucky/Louisville"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("dearborn", "ohio"), "county"),
          "tz"] <- "America/New_York"
county_df[county_df$state == "IN" &
            county_df$county %in% paste(c("gibson", "jasper", "lake",
                                          "laporte", "newton", "perry",
                                          "porter", "posey", "spencer",
                                          "starke", "vanderburgh",
                                          "warrick"), "county"), "tz"] <- "America/Chicago"
county_df[county_df$state == "KS" &
            county_df$county %in% paste(c("greeley", "hamilton",
                                          "sherman", "wallace"),
                                        "county"), "tz"] <- "America/Denver"
county_df[county_df$state == "KY" &
            county_df$county %in% paste(c("adair", "allen", "ballard", "barren",
                                          "breckinridge", "butler", "caldwell",
                                          "calloway", "carlisle", "christian",
                                          "clinton", "crittenden", "cumberland",
                                          "daviess", "edmonson", "fulton", "graves",
                                          "grayson", "green", "hancock", "hart",
                                          "henderson", "hickman", "hopkins",
                                          "livingston", "logan", "lyon", "marshall",
                                          "mccracken", "mclean", "metcalfe", "monroe",
                                          "muhlenberg", "ohio", "russell", "simpson",
                                          "todd", "trigg", "union", "warren", "webster"),
                                        "county"), "tz"] <- "America/Chicago"
county_df[county_df$state == "KY" &
            county_df$county %in% paste(c("jefferson"),
                                        "county"), "tz"] <- "America/Kentucky/Louisville"
county_df[county_df$state == "KY" &
            county_df$county %in% paste(c("wayne"),
                                        "county"), "tz"] <- "America/Kentucky/Monticello"
county_df[county_df$state == "MI" &
            county_df$county %in% paste(c("dickinson", "gogebic", "iron",
                                          "menominee"),
                                        "county"), "tz"] <- "America/Menominee"
county_df[county_df$state == "NE" &
            county_df$county %in% paste(c("arthur", "banner", "box butte",
                                          "chase", "cherry", "cheyenne", "dawes", "deuel",
                                          "dundy", "garden", "grant", "hooker",
                                          "keith", "kimball", "morrill",
                                          "perkins", "scotts bluff", "sheridan",
                                          "sioux"),
                                        "county"), "tz"] <- "America/Denver"
county_df[county_df$state == "ND" &
            county_df$county %in% paste(c("adams", "billings", "bowman", "dunn",
                                          "golden valley", "grant", "hettinger",
                                          "slope", "stark"),
                                        "county"), "tz"] <- "America/Denver"
county_df[county_df$state == "ND" &
            county_df$county %in% paste(c("oliver"),
                                        "county"), "tz"] <- "America/North_Dakota/Center"
county_df[county_df$state == "ND" &
            county_df$county %in% paste(c("morton"),
                                        "county"), "tz"] <- "America/North_Dakota/New_Salem"
county_df[county_df$state == "ND" &
            county_df$county %in% paste(c("mercer"),
                                        "county"), "tz"] <- "America/North_Dakota/Beulah"
county_df[county_df$state == "OR" &
            county_df$county %in% paste(c("malheur"),
                                        "county"), "tz"] <- "America/Boise"
county_df[county_df$state == "SD" &
            county_df$county %in% paste(c("bennett", "butte", "corson", "custer",
                                          "dewey", "fall river", "haakon", "harding",
                                          "jackson", "lawrence", "meade", "pennington",
                                          "perkins", "shannon", "stanley", "ziebach"),
                                        "county"), "tz"] <- "America/Denver"
county_df[county_df$state == "TN" &
            county_df$county %in% paste(c("anderson", "blount", "bradley",
                                          "campbell", "carter", "claiborne",
                                          "cocke", "grainger", "greene",
                                          "hamblen", "hamilton", "hancock",
                                          "hawkins", "jefferson", "johnson",
                                          "knox", "loudon", "mcminn", "meigs",
                                          "monroe", "morgan", "polk", "rhea",
                                          "roane", "scott", "sevier", "sullivan",
                                          "unicoi", "union", "washington"),
                                        "county"), "tz"] <- "America/New_York"
county_df[county_df$state == "TX" &
            county_df$county %in% paste(c("el paso", "hudspeth"),
                                        "county"), "tz"] <- "America/Denver"
county_tzs <- select(county_df, fips, tz, county, state)
library(devtools)
use_data(county_tzs, overwrite = TRUE)


R/add_local_time.R
#' Add local time to dataset
#'
#' This function inputs a dataframe of observations for US counties that
#' includes a column with date-time expressed in Coordinated Universal Time
#' (UTC), as well as a single value or vector of FIPS code(s) for the
#' county associated with each observation. It will return the
#' original dataframe with columns added for local date-time, local date, and,
#' if specified, local time zone.
#'
#' @param df A dataframe of observations that includes a column with a
#'    date-time object in Coordinated Universal Time (UTC) (see the documentation
#'    for the \code{\link{calc_local_time}} function to see requirements for
#'    the format of this UTC date-time column)
#' @param datetime_colname A character string giving the column name for the
#'    column that gives date-time in UTC in the input dataframe.
#' @inheritParams calc_local_time
#'
#' @return This function returns the dataframe input to the function, with
#'    columns added with local date-time, local date, and, if specified, local
#'    time zone.
#'
#' @examples
#' ex_df <- data.frame(datetime = c("1999-01-01 08:00", "1999-01-01 09:00",
#'                                  "1999-01-01 10:00"),
#'                     fips = c("36061", "17031", "06037"))
#' add_local_time(df = ex_df, fips = ex_df$fips,
#'                datetime_colname = "datetime")
#'
#' ex_df <- data.frame(datetime = c("1999-01-01 08:00", "1999-01-01 09:00",
#'                                  "1999-01-01 10:00"))
#' add_local_time(df = ex_df, fips = "36061", datetime_colname = "datetime")
#'
#' @export
add_local_time <- function(df, fips, datetime_colname, include_tz = TRUE){
  date_time <- df[ , datetime_colname]
  local_time <- calc_local_time(date_time = date_time, fips = fips,
                                include_tz = include_tz)
  out <- cbind(df, local_time)
  return(out)
}

#' Calculate local time from UTC for US counties
#'
#' This function inputs date-time values in Coordinated Universal Time (UTC;
#' also known as Zulu Time), along with a vector with county Federal Information
#' Processing Standard (FIPS) codes, and calculates the local date-time as
#' well as the local date based on th UTC date-time.
#'
#' This function inputs date-time values in Coordinated Universal Time (UTC;
#' also known as Zulu Time), as well as a single value or vector of FIPS code(s) for the
#' county associated with each observation. It returns a dataframe with
#' columns for local date-time, local date, and, if specified, local time zone.
#'
#' @param date_time The vector of the date-time of each observation in
#'    Coordinated Universal Time (UTC). This vector can either have a
#'    \code{POSIXct} class or be a character string, with date-time given
#'    as four-digit year, two-digit month, two-digit day, two-digit hour,
#'    and two-digit minutes (with hours based on a 24-hour system). Examples
#'    of acceptable formats include, for the example of 1:00 PM Jan. 2 1999,
#'    "199901021300", "1999-01-02 13:00", and "1999/01/02 13:00".
#' @param fips A character vector giving the 5-digit FIPS code of the county
#'    associated with each observation. This can be either a string of length 1,
#'    if all observations come from the same county, or a vector as long as the
#'    \code{date_time} vector, if different observations come from different
#'    counties.
#' @param include_tz A TRUE / FALSE value specifying whether to include a
#'    column with the local time zone (\code{local_tz}) in the final output.
#'
#' @return This function returns a dataframe with columns for local date-time,
#'    local date, and, if specified, local time zone.
#'
#' @note The local time is given as a character
#'    string, rather than a POSIXct object, so that it can have different time
#'    zones for different counties within the same dataframe (i.e., if there
#'    are two counties in the dataframe that are in different time zones).
#'
#' @examples
#' calc_local_time(date_time = "1999-01-01 08:00", fips = "36061")
#'
#' ex_datetime <- c("1999-01-01 08:00", "1999-01-01 09:00",
#'                  "1999-01-01 10:00")
#' calc_local_time(date_time = ex_datetime, fips = "36061")
#'
#' ex_datetime <- c("1999-01-01 08:00", "1999-01-01 09:00",
#'                  "1999-01-01 10:00")
#' ex_fips <- c("36061", "17031", "06037")
#' calc_local_time(date_time = ex_datetime, fips = ex_fips)
#'
#' @importFrom dplyr %>%
#'
#' @export
calc_local_time <- function(date_time, fips, include_tz = TRUE){
  fips <- as.numeric(as.character(fips))

  wrong_fips <- fips[!(as.numeric(fips) %in% countytimezones::county_tzs$fips)]
  if (length(wrong_fips) > 0){
    warning(paste("The following FIPS did not match values in our dataset:",
                  paste(wrong_fips, collapse = ", ")))
    fips <- fips[(fips %in% countytimezones::county_tzs$fips)]
   }

  # Convert date-time to POSIXct class if it's not already
  if(!("POSIXct" %in% class(date_time))){
    safe_ymd_hm <- purrr::safely(lubridate::ymd_hm)
    date_time <- safe_ymd_hm(date_time)
    if(!is.null(date_time$error)){
      stop("The `date_time` must be in a format like `1999-01-01 08:00`.")
    }
    date_time <- date_time$result
  }

  if(include_tz){
    dots <- c("local_time", "local_date", "local_tz")
  } else {
    dots <- c("local_time", "local_date")
  }

  df <- data.frame(date_time, fips) %>%
    dplyr::mutate_(fips = ~ as.numeric(as.character(fips))) %>%
    dplyr::left_join(countytimezones::county_tzs, by = "fips") %>%
    dplyr::rename_(local_tz = ~ tz) %>%
    dplyr::mutate_(local_time = ~ mapply(calc_single_datetime,
                                         date_time, tz = local_tz),
                   local_time = ~ lubridate::ymd_hms(local_time),
                   local_date = ~ format(local_time, "%Y-%m-%d"),
                   local_time = ~ format(local_time, "%Y-%m-%d %H:%M")) %>%
    dplyr::select_(.dots = dots)
  return(df)
}

#' Convert UTC to local time for a single observation
#'
#' This function calculated the local date-time for an observation based on a
#' date-time in Coordinated Universal Time (UTC). The function provides a
#' wrapper for the \code{with_tz} function form the \code{lubridate} package.
#' It converts output from the \code{with_tz} function to a character vector
#' so other functions in this package can be applied without error to
#' with a dataframe with observations from multiple time zones to local time.
#'
#' @param datetime A POSIXct object of length one expressed in Coordinated
#'    Universal Time (UTC)
#' @param tz A character string giving the local time zone based on the
#'    Olson/IANA time zone names
#'
#' @return A character string giving the date-time in the local time zone
#'
#' @examples
#' utc_time <- as.POSIXct("1999-09-15 14:30:00", tz = "UTC")
#' local_time <- calc_single_datetime(utc_time, tz = "US/Eastern")
#'
#' @export
calc_single_datetime <- function(datetime, tz){
  local_time <- lubridate::with_tz(datetime, tzone = tz)
  local_time <- format(local_time, format = "%Y%m%d%H%M%S")
  return(local_time)
}


